src/sp/it/pl/gui/objects/contextmenu/ImprovedContextMenu.kt

```diff
@@ -103,34 +110,53 @@ class ContextMenuItemSuppliers {
     private val mMany = ClassListMap<ItemsSupply> { fail() }
 
     @Suppress("UNCHECKED_CAST")
-    fun <T: Any> add(type: Class<T>, items: (ImprovedContextMenu<*>, T) -> Sequence<MenuItem>) {
-        mSingle.accumulate(type, items as ItemsSupply)
+    fun <T: Any> add(type: Class<T>, items: ContextMenuBuilder<T>.() -> Unit) {
+        mSingle.accumulate(type) { menu, item ->
+            ContextMenuBuilder(menu, item as T).also { items(it) }.asSequence()
+        }
     }
 
     @Suppress("UNCHECKED_CAST")
-    fun <T: Any> addMany(type: Class<T>, items: (ImprovedContextMenu<*>, Collection<T>) -> Sequence<MenuItem>) {
-        mMany.accumulate(type, items as ItemsSupply)
+    fun <T: Any> addMany(type: Class<T>, items: ContextMenuBuilder<Collection<T>>.() -> Unit) {
+        mMany.accumulate(type) { menu, item ->
+            ContextMenuBuilder(menu, item as Collection<T>).also { items(it) }.asSequence()
+        }
     }
 
-    inline fun <reified T: Any> add(noinline items: (ImprovedContextMenu<*>, T) -> Sequence<MenuItem>) {
+    inline fun <reified T: Any> add(noinline items: ContextMenuBuilder<T>.() -> Unit) {
         add(T::class.java, items)
     }
 
-    inline fun <reified T: Any> addMany(noinline items: (ImprovedContextMenu<*>, Collection<T>) -> Sequence<MenuItem>) {
+    inline fun <reified T: Any> addMany(noinline items: ContextMenuBuilder<Collection<T>>.() -> Unit) {
         addMany(T::class.java, items)
     }
 
     operator fun get(contextMenu: ImprovedContextMenu<*>, value: Any?): Sequence<MenuItem> {
+        logger.trace { "Generating contextmenu for $value" }
         val items1 = mSingle.getElementsOfSuperV(value?.javaClass ?: Void::class.java).asSequence()
-                .map { it(contextMenu, value) }
-                .flatMap { it }
+                .flatMap { it(contextMenu, value) }
         val itemsN = if (value is Collection<*>) {
             mMany.getElementsOfSuperV(value.getElementType()).asSequence()
-                    .map { it(contextMenu, contextMenu.value) }
-                    .flatMap { it }
+                    .flatMap { it(contextMenu, contextMenu.value) }
         } else {
             sequenceOf()
         }
         return items1+itemsN
     }
+
+}
+
+class ContextMenuBuilder<T>(val contextMenu: ImprovedContextMenu<*>, val selected: T): ArrayList<MenuItem>() {
+
+    fun <T: MenuItem> T.add() = also { add(it) }
+
+    fun item(text: String, handler: (ActionEvent) -> Unit): MenuItem =
+            menuItem(text, handler).add()
+
+    fun menu(text: String, graphic: Node? = null, items: Menu.() -> Unit) =
+            sp.it.pl.util.graphics.menu(text, graphic, items).add()
+
+    fun menu(text: String, items: Sequence<MenuItem> = seqOf()) =
```

Comment:
as extension methods to what?

Reply:
The builder. Basically instead of the fun being a member method declared in the class, it is defined next to it as top level extension function. I would never do this except this one case with dsl.

Reply:
Unifies extensions with the methods the builder is inherently equipped with. IDEA has special visuals for extension methods and its easy on the eyes if the whole dsl follows the same pattern. I do agree this may be going too far. In my opinion it helps readability because the structural code of the menu (menu {}, item {}) will pop out. If some methods are extension and some not, the overall structure is lost and even worse partially. Some of the structure is highlighted and some is not. It is very confusing and makes one question why one menu() is standard method (which developer would expect to be constructor) while the other is an extension. With builders it is sometimes better for everything to be an extension so developer can distinguish structural and functional code.